doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+thing! Pushing to achieve full coverage of annotated code can lead to low signal-to-noise and sometimes is not practical
+or impossible given the expressiveness of Python as a language. So, in practice, what should you aim to type?
+
+1) Add type hints to publicly exposed APIs in the client library. Type hints get shipped with our client libraries and
+   provide benefit to our users. To ensure that type hints do get shipped in your library and can be checked by a static
+   type checker, follow the steps
+   per [PEP 561](https://mypy.readthedocs.io/en/stable/installed_packages.html#creating-pep-561-compatible-packages)
+   below:
+
+    - add an empty `py.typed` file to your package directory. E.g. `.../sdk/azure-core/azure/core/py.typed`
+    - include the `py.typed` file in the
+      MANIFEST.in ([example](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/MANIFEST.in))
+
+2) Add type hints anywhere in the source code where unit tests are worth writing. Consider typing/mypy as "free" tests
+   for your library so focusing typing on high density/important areas of the code helps in detecting bugs.
+
+### Types usable in annotations
+
+Almost anything can be used as a type in annotations.
+
+1) Basic types like `int`, `str`, `bytes`, `float`, `bool`
+2) Classes or abstract base classes - whether user-defined, from the standard library
+   like [collections](https://docs.python.org/3/library/collections.html)
+   or [collections.abc](https://docs.python.org/3/library/collections.abc.html), or external packages
+3) Types from the [typing](https://docs.python.org/3/library/typing.html)
+   or [typing_extensions](https://github.com/python/typing/tree/master/typing_extensions) modules
+4) Built-in generic types, like `list` or `dict`*.
+   > *Note: Support for generic types was not added until Python 3.9 (or 3.7+ with `from __future__ import annotations` import). This lets you pass in generic `list[str]` as a type hint rather than `typing.List[str]`.
+
+Here are a few considerations and notes on Python version support:
+
+- With [PEP585](https://peps.python.org/pep-0585/) and [PEP563](https://peps.python.org/pep-0563/), importing certain
+  generic collection types from `typing` has been [deprecated](https://peps.python.org/pep-0585/#implementation) in
+  favor for using the types in `collections.abc` or generic type hints (like `list`) from the standard library. At the
+  time of writing, the Python SDK supports 3.6+ and so we must still import these types from `typing` or check the
+  Python version at runtime before importing (Note that the `collections.abc` types did not have [] notation added until
+  3.9)
+
+```python
+import sys
+
+if sys.version_info < (3, 9):
+    from typing import Sequence
+else:
+    from collections.abc import Sequence
+```
+
+- The `typing-extensions` library backports types that are introduced only in later versions of Python (e.g. `Protocol`
```

Comment:
We should have a note somewhere that typing-extensions is doing a try/except, and therefore people should always import from type-extensions (and not try/except from typing/type-extensions)